File: solutions/clojure/perfect-numbers/1/src/perfect_numbers.clj
(ns perfect-numbers)

(defn aliquot-sum [number]
  (reduce + (filter #(zero? (mod number %))
                    (range 1 number))))

(defn classify [number]
  (when (< number 0)
    (throw (IllegalArgumentException.)))
  (let [sum (aliquot-sum number)]
    (cond
      (< sum number) :deficient
      (> sum number) :abundant
      :else :perfect)))

File: solutions/fortran/perfect-numbers/1/perfect_numbers.f90

module perfect_numbers
   implicit none

contains

   character(len=9) function classify(num)
      integer, intent(in) :: num
      integer :: i, sum
      sum = 0
      if (num < 1) then
         classify = "ERROR"
      else
         do i = 1, num - 1
            if (mod(num, i) == 0) then
               sum = sum + i
            end if
         end do
         if (sum == num) then
            classify = "perfect"
         else if (sum > num) then
            classify = "abundant"
         else
            classify = "deficient"
         end if
      end if
   end function

end module
